# Usando la validacion cruzada

En el notebook anterior, construimos un modelo de Bosques Aleatorios y realizamos una búsqueda manual de los hiperparámetros óptimos, como el número de árboles y la profundidad máxima. Si bien este enfoque nos dio un modelo con un sólido rendimiento del 82%, es crucial validar nuestros hallazgos de una manera más robusta.

En este notebook, llevaremos la optimización de nuestro modelo al siguiente nivel. Utilizaremos validación cruzada junto con GridSearchCV para encontrar la combinación de hiperparámetros que garantice el mejor rendimiento de manera sistemática y confiable.

El objetivo es confirmar si los valores que elegimos manualmente eran de hecho los óptimos y, al mismo tiempo, obtener un resultado que minimice el riesgo de sobreajuste al probar el modelo en múltiples divisiones de datos.

In [1]:
# Importamos las librerías necesarias
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import f1_score

# Cargamos el dataset y separamos los datos
df_model_wine = pd.read_csv("dataset/wine_ready.csv")
X = df_model_wine.drop('quality_binary', axis=1)
y = df_model_wine['quality_binary']

# Dividimos los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

# Definimos el modelo base
rf_model = RandomForestClassifier(random_state=42)

# Definimos la "rejilla" de parámetros a probar
# Incluimos los rangos alrededor de los valores óptimos que ya encontraste
param_grid = {
    'n_estimators': [160, 170, 180],
    'max_depth': [18, 19, 20],
    'min_samples_leaf': [1, 2],
    'max_features': ['sqrt', 'log2']  # Nuevo parámetro a probar
}

# Creamos el objeto GridSearchCV
grid_search = GridSearchCV(
    estimator=rf_model,
    param_grid=param_grid,
    scoring='f1_weighted',  # Usamos el f1-score como métrica
    cv=5,                  # Usamos 5-fold cross-validation
    n_jobs=-1,             # Usa todos los procesadores disponibles
    verbose=1              # Muestra el progreso
)

# Ejecutamos la búsqueda exhaustiva
grid_search.fit(X_train, y_train)

# Imprimimos los mejores parámetros y el mejor score
print("Mejores parámetros encontrados:")
print(grid_search.best_params_)
print(f"\nMejor F1-Score (promedio de 5-fold CV): {grid_search.best_score_:.4f}")

Fitting 5 folds for each of 36 candidates, totalling 180 fits
Mejores parámetros encontrados:
{'max_depth': 20, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'n_estimators': 170}

Mejor F1-Score (promedio de 5-fold CV): 0.8147


#### Conclusiones

Hemos completado la fase de optimización de nuestro modelo de Bosques Aleatorios utilizando la validación cruzada. La búsqueda exhaustiva de GridSearchCV ha arrojado la siguiente combinación de parámetros como la más óptima para nuestro conjunto de datos:

1. n_estimators: 170

2. max_depth: 20

3. min_samples_leaf: 1

4. max_features: sqrt

El resultado final, con un F1-Score promedio de 0.8147, confirma que los parámetros encontrados en la búsqueda manual eran extraordinariamente cercanos a los verdaderos valores óptimos. Este hallazgo valida la solidez de nuestro análisis exploratorio y la metodología de optimización.

Ahora tenemos un modelo final, cuyos hiperparámetros no han sido elegidos al azar, sino que han sido validados rigurosamente. Este modelo es robusto, confiable y está listo para ser utilizado para la predicción de nuevos datos.